### ISSUE TITLE
Ping timeout and reconnect result in panic: close of closed channel 

### ISSUE THREAD

**Issue by lidel** [thoj/go-ircevent#65]
When connection drops and ping timeout triggers reconnect [this line](https://github.com/thoj/go-ircevent/blob/f69e2d34447f899c9fd0984c8e3cfbb1271e0381/irc.go#L320) tries to close already closed channel:

```
2015/08/24 23:28:04 Lag: 121.405416mss
2015/08/24 23:30:24 <-- PING :asimov.freenode.net
2015/08/24 23:30:24 --> PONG :asimov.freenode.net
2015/08/24 23:32:39 <-- PING :asimov.freenode.net
2015/08/24 23:32:39 --> PONG :asimov.freenode.net
2015/08/24 23:34:54 <-- :meowkov!~meowkov@host.foo QUIT :Ping timeout: 255 seconds
2015/08/24 23:34:54 <-- ERROR :Closing Link: host.foo (Ping timeout: 255 seconds)
2015/08/24 23:34:54 Error, disconnected: EOF
panic: close of closed channel

goroutine 1 [running]:
github.com/thoj/go-ircevent.(*Connection).Reconnect(0xc8200905a0, 0x0, 0x0)
        /go/src/github.com/thoj/go-ircevent/irc.go:362 +0x38
github.com/thoj/go-ircevent.(*Connection).Loop(0xc8200905a0)
        /go/src/github.com/thoj/go-ircevent/irc.go:199 +0x1ac
main.ircLoop()
        /meowkov/meowkov.go:222 +0x28d
main.main()
        /meowkov/meowkov.go:135 +0x8f

goroutine 17 [syscall, 3318 minutes, locked to thread]:
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1696 +0x1

goroutine 4 [chan send]:
github.com/thoj/go-ircevent.(*Connection).readLoop(0xc8200905a0)
        /go/src/github.com/thoj/go-ircevent/irc.go:69 +0x38c
created by github.com/thoj/go-ircevent.(*Connection).Connect
        /go/src/github.com/thoj/go-ircevent/irc.go:425 +0xd22

goroutine 18678 [semacquire]:
sync.runtime_Semacquire(0xc8200905ac)
        /usr/local/go/src/runtime/sema.go:43 +0x26
sync.(*WaitGroup).Wait(0xc8200905a0)
        /usr/local/go/src/sync/waitgroup.go:126 +0xb4
github.com/thoj/go-ircevent.(*Connection).Disconnect(0xc8200905a0)
        /go/src/github.com/thoj/go-ircevent/irc.go:352 +0x11e
github.com/thoj/go-ircevent.(*Connection).setupCallbacks.func1(0xc8200a2180)
        /go/src/github.com/thoj/go-ircevent/irc_callback.go:145 +0x20
created by github.com/thoj/go-ircevent.(*Connection).RunCallbacks
        /go/src/github.com/thoj/go-ircevent/irc_callback.go:123 +0x524
```

I did not dig into this yet, perhaps @dainis (who authored that part in https://github.com/thoj/go-ircevent/commit/f69e2d34447f899c9fd0984c8e3cfbb1271e0381) sees a quick fix? 

It seems that this is a regression, as the same issue was fixed in 2014 in https://github.com/thoj/go-ircevent/pull/33

**Comment by thoj** [comment]
Hm I really don't understand how this can happen because the channel is only closed in one place. (In Reconnect). Was the channel initialized at all?

**Comment by lidel** [comment]
I think so. [The bot](https://github.com/lidel/meowkov) was connected to the server for days. 
Everything worked fine.  Then ping timeout happened and triggered this panic..

**Comment by dainis** [comment]
Error mesasge is triggering call to disconnect

Correct me if I'm wrong but this default Error message handling approach has side effects, it would remove all callbacks, and then when it reconnects there won't be any as they are setup when IRC object is created

**Comment by lidel** [comment]
Hm.. current version simply dies on timeout (there is no reconnect):

```
2015/11/28 05:40:08 --> PING 1448689208550955025
2015/11/28 05:41:08 --> PING 1448689268558019719
2015/11/28 05:42:08 --> PING 1448689328551226945
2015/11/28 05:43:08 --> PING 1448689388551876976
2015/11/28 05:44:08 --> PING 1448689448558236717
2015/11/28 05:45:08 --> PING 1448689508551374293
2015/11/28 05:46:08 --> PING 1448689568550824439
2015/11/28 05:47:08 --> PING 1448689628552612945
2015/11/28 05:47:32 Error, disconnected: read tcp 172.17.0.2:59441->130.239.18.119:7000: i/o timeout
```

**Comment by peted27** [comment]
ive created a pull request to fix reconnection issues, #69 - it may fix the issues youve been having

**Comment by lidel** [comment]
Thank you @peted27, it indeed fixed reconnect logic :+1: 

````
2015/12/06 16:39:20 Error, disconnected: EOF
2015/12/06 16:39:20 Error while reconnecting: dial tcp 37.187.XXX.XXX:6665: getsockopt: connection refused
2015/12/06 16:40:20 Error while reconnecting: dial tcp 37.187.XXX.XXX:6665: getsockopt: connection refused
2015/12/06 16:41:20 Connected to node.foo:6665 (37.187.XXX.XXX:6665)
```
